subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestWorker.java

```diff
@@ -42,18 +42,25 @@
 
 import java.io.Serializable;
 import java.security.AccessControlException;
-import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
 
 public class TestWorker implements Action<WorkerProcessContext>, RemoteTestClassProcessor, Serializable, Stoppable {
     private static final Logger LOGGER = LoggerFactory.getLogger(TestWorker.class);
     public static final String WORKER_ID_SYS_PROPERTY = "org.gradle.test.worker";
     public static final String WORKER_TMPDIR_SYS_PROPERTY = "org.gradle.internal.worker.tmpdir";
 
     private final WorkerTestClassProcessorFactory factory;
-    private CountDownLatch completed;
+    private LinkedBlockingQueue<TestClassRunInfo> toRun;
```

Comment:
We could replace the unbounded `LinkedBlockingQueue` with `ArrayBlockingQueue(capacity)`. This would reduce GC pressure a bit, and also cause `processTestClass(TestClassRunInfo testClass)` to sometimes block until the queue has space. Right now `processTestClass` is always able to complete immediately because the unbounded queue can hold the backlog.

Reply:
I think an unbounded queue is just asking for trouble.  _Someone_ will be running a test suite with a billion small tests and a small worker heap size and they won't have any way to alleviate the issue without just adding heap, which may have other implications.  This should be bounded in some way.

Reply:
I think that's smart. I believe the code before this change was equivalent to `ArrayBlockingQueue(1)`. Probably you get a *little* bit of performance help by letting them buffer a bit. Sizing it is a bit arbitrary, but I'll try `128` with this rough reasoning:

`1KiB` allocator block / `8 byte` pointer on 64-bit machines = `128`